src/schemes/boundary/open_boundary/system.jl

```diff
@@ -482,6 +345,30 @@ function wrap_reference_function(constant_vector_, ::Val{NDIMS}) where {NDIMS}
     return constant_vector(coords, t) = SVector{NDIMS}(constant_vector_)
 end
 
+function reference_value(value::Function, quantity, system, particle, position, t)
+    return value(position, t)
+end
+
+# These methods are used when extrapolating quantities from the domain
+# instead of using the method of characteristics
+function reference_value(value::Nothing, quantity::Vector, system, particle, position, t)
+    return quantity[particle]
+end
+
+function reference_value(value::Nothing, quantity::PtrArray, system, particle, position, t)
```

Comment:
This is relying on used data types, which might change in the future, and which are also different on GPUs. Please add an extra function `reference_velocity` or so, instead of dispatching on this type.